File: bindings/dotnet/AgentStateGraph/Taint.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;
using AgentStateGraph.Interop;

namespace AgentStateGraph;

// =============================================================================
// Enums (0.7.75 §7 — taint / quarantine / watch)
// =============================================================================

/// <summary>
/// Discriminator between the three taint kinds.
/// Matches Rust <c>agentstategraph_taint::TaintKind</c>.
/// </summary>
public enum TaintKind
{
    Taint,
    Quarantine,
    Watch,
}

/// <summary>
/// Pre-commit hook behavior for a taint. Matches Rust
/// <c>agentstategraph_taint::TaintEffect</c>.
/// </summary>
public enum TaintEffect
{
    /// <summary>Attach a warning but allow the write.</summary>
    Warn,
    /// <summary>Reject the write.</summary>
    Block,
    /// <summary>Require confidence &gt;= 0.9; reject lower confidence writes.</summary>
    Review,
    /// <summary>Allow the write; flag the path for query / search filtering.</summary>
    Isolate,
    /// <summary>Watch-only — purely advisory.</summary>
    Advisory,
}

/// <summary>Advisory severity for a taint — does not change pre-commit semantics.</summary>
public enum TaintSeverity
{
    Low,
    Medium,
    High,
    Critical,
}

/// <summary>Direction a watch threshold fires in.</summary>
public enum WatchDirection
{
    Above,
    Below,
}

// =============================================================================
// Records — wire shapes from crates/agentstategraph-taint/src/types.rs
// =============================================================================

/// <summary>
/// A taint / quarantine / watch record as emitted by
/// <see cref="Repository.ListTaints"/> and the
/// <c>agentstategraph_check_taint</c> envelope.
/// </summary>
public sealed record Taint(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("kind")] TaintKind Kind,
    [property: JsonPropertyName("effect")] TaintEffect Effect,
    [property: JsonPropertyName("severity")] TaintSeverity Severity,
    [property: JsonPropertyName("reason")] string Reason,
    [property: JsonPropertyName("agent_id")] string AgentId,
    [property: JsonPropertyName("commit_id")] string CommitId,
    [property: JsonPropertyName("created_at")] DateTimeOffset CreatedAt,
    [property: JsonPropertyName("propagate")] bool Propagate,
    [property: JsonPropertyName("expires_at")] DateTimeOffset? ExpiresAt = null,
    [property: JsonPropertyName("resolved_at")] DateTimeOffset? ResolvedAt = null,
    [property: JsonPropertyName("resolved_by")] string? ResolvedBy = null,
    [property: JsonPropertyName("resolved_reason")] string? ResolvedReason = null,
    [property: JsonPropertyName("resolved_proof")] string? ResolvedProof = null,
    /// <summary>
    /// Kind-specific metadata. Typed as <see cref="JsonElement"/> so callers
    /// can inspect the shape without the binding committing to a schema
    /// (matches the <see cref="Task.Payload"/> pattern).
    /// </summary>
    [property: JsonPropertyName("metadata")] IReadOnlyDictionary<string, JsonElement>? Metadata = null);

/// <summary>
/// Result of <see cref="Repository.CheckTaint"/>. Mirrors Rust
/// <c>agentstategraph_taint::TaintCheck</c>.
/// </summary>
public sealed record TaintCheck(
    [property: JsonPropertyName("tainted")] bool Tainted,
    [property: JsonPropertyName("quarantined")] bool Quarantined,
    [property: JsonPropertyName("watched")] bool Watched,
    [property: JsonPropertyName("can_write")] bool CanWrite,
    [property: JsonPropertyName("required_confidence")] double RequiredConfidence,
    [property: JsonPropertyName("isolated")] bool Isolated,
    [property: JsonPropertyName("taints")] IReadOnlyList<Taint>? Taints = null,
    [property: JsonPropertyName("quarantines")] IReadOnlyList<Taint>? Quarantines = null,
    [property: JsonPropertyName("watches")] IReadOnlyList<Taint>? Watches = null,
    [property: JsonPropertyName("authorized_agents")] IReadOnlyList<string>? AuthorizedAgents = null);

// =============================================================================
// Parameter records for the Repository surface
// =============================================================================

/// <summary>Parameters for <see cref="Repository.Taint"/>.</summary>
public sealed record TaintParams(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("effect")] TaintEffect Effect,
    [property: JsonPropertyName("reason")] string Reason,
    [property: JsonPropertyName("agent_id")] string AgentId,
    [property: JsonPropertyName("severity")] TaintSeverity Severity = TaintSeverity.Medium,
    [property: JsonPropertyName("expires")] DateTimeOffset? ExpiresAt = null,
    [property: JsonPropertyName("propagate")] bool Propagate = true,
    [property: JsonPropertyName("metadata")] IReadOnlyDictionary<string, JsonElement>? Metadata = null);

/// <summary>Parameters for <see cref="Repository.Untaint"/> / <see cref="Repository.Unquarantine"/>.</summary>
public sealed record UntaintParams(
    [property: JsonPropertyName("reason")] string Reason,
    [property: JsonPropertyName("agent_id")] string AgentId,
    [property: JsonPropertyName("proof")] string? Proof = null);

/// <summary>Parameters for <see cref="Repository.Quarantine"/>.</summary>
public sealed record QuarantineParams(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("reason")] string Reason,
    [property: JsonPropertyName("authorized_agents")] IReadOnlyList<string> AuthorizedAgents,
    [property: JsonPropertyName("agent_id")] string AgentId,
    [property: JsonPropertyName("severity")] TaintSeverity Severity = TaintSeverity.Medium,
    [property: JsonPropertyName("expires")] DateTimeOffset? ExpiresAt = null,
    [property: JsonPropertyName("propagate")] bool Propagate = true);

/// <summary>Parameters for <see cref="Repository.Watch"/>.</summary>
public sealed record WatchParams(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("reason")] string Reason,
    [property: JsonPropertyName("agent_id")] string AgentId,
    [property: JsonPropertyName("metric")] string? Metric = null,
    [property: JsonPropertyName("threshold")] double? Threshold = null,
    [property: JsonPropertyName("direction")] WatchDirection Direction = WatchDirection.Above,
    [property: JsonPropertyName("check_interval_secs")] ulong? CheckIntervalSecs = null,
    [property: JsonPropertyName("expires")] DateTimeOffset? ExpiresAt = null,
    [property: JsonPropertyName("severity")] TaintSeverity Severity = TaintSeverity.Medium,
    [property: JsonPropertyName("propagate")] bool Propagate = true);

/// <summary>Parameters for <see cref="Repository.Unwatch"/>.</summary>
public sealed record UnwatchParams(
    [property: JsonPropertyName("agent_id")] string AgentId,
    [property: JsonPropertyName("reason")] string? Reason = null);

// =============================================================================
// Repository instance API (§7 pass-through)
// =============================================================================

/// <summary>
/// Taint / quarantine / watch methods on <see cref="Repository"/>. Kept in a
/// partial class so the core repository file isn't buried in §7 surface.
/// </summary>
public sealed partial class Repository
{
    /// <summary>
    /// Applies a taint to <paramref name="path"/>. Returns the new taint id.
    /// </summary>
    public string Taint(string path, TaintParams p, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(p);
        var ptr = NativeMethods.agentstategraph_taint_apply(
            _handle.DangerousGetHandle(), refName, path, Json.Serialize(p));
        return ParseIdEnvelope(Strings.ConsumeUtf8(ptr), "taint_apply");
    }

    /// <summary>Removes the taint named <paramref name="name"/> from <paramref name="path"/>.</summary>
    public void Untaint(string path, string name, UntaintParams p, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(p);
        var body = BuildRemoveParams(name, p.Reason, p.AgentId, p.Proof);
        var ptr = NativeMethods.agentstategraph_taint_remove(
            _handle.DangerousGetHandle(), refName, path, body);
        Json.ThrowIfError(Strings.ConsumeUtf8(ptr), "taint_remove");
    }

    /// <summary>Applies a quarantine. Returns the new taint id.</summary>
    public string Quarantine(string path, QuarantineParams p, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(p);
        var ptr = NativeMethods.agentstategraph_quarantine_apply(
            _handle.DangerousGetHandle(), refName, path, Json.Serialize(p));
        return ParseIdEnvelope(Strings.ConsumeUtf8(ptr), "quarantine_apply");
    }

    /// <summary>Releases a quarantine.</summary>
    public void Unquarantine(string path, string name, UntaintParams p, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(p);
        var body = BuildRemoveParams(name, p.Reason, p.AgentId, p.Proof);
        var ptr = NativeMethods.agentstategraph_quarantine_release(
            _handle.DangerousGetHandle(), refName, path, body);
        Json.ThrowIfError(Strings.ConsumeUtf8(ptr), "quarantine_release");
    }

    /// <summary>Applies a watch. Returns the new taint id.</summary>
    public string Watch(string path, WatchParams p, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(p);
        var ptr = NativeMethods.agentstategraph_watch_apply(
            _handle.DangerousGetHandle(), refName, path, Json.Serialize(p));
        return ParseIdEnvelope(Strings.ConsumeUtf8(ptr), "watch_apply");
    }

    /// <summary>Removes a watch.</summary>
    public void Unwatch(string path, string name, UnwatchParams p, string refName = DefaultRef)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(p);
        // Build a body that includes `name` (required by the FFI) plus the
        // serialized unwatch params (reason + agent_id).
        using var baseDoc = JsonDocument.Parse(Json.Serialize(p));
        var body = MergeWithName(baseDoc.RootElement, name);
        var ptr = NativeMethods.agentstategraph_watch_remove(
            _handle.DangerousGetHandle(), refName, path, body);
        Json.ThrowIfError(Strings.ConsumeUtf8(ptr), "watch_remove");
    }

    /// <summary>
    /// Lists taints / quarantines / watches. Filters by <paramref name="pathPrefix"/>
    /// and / or <paramref name="kind"/> when provided.
    /// </summary>
    public IReadOnlyList<Taint> ListTaints(
        string? pathPrefix = null,
        TaintKind? kind = null,
        bool includeResolved = false)
    {
        ThrowIfDisposed();
        var kindStr = kind switch
        {
            TaintKind.Taint => "taint",
            TaintKind.Quarantine => "quarantine",
            TaintKind.Watch => "watch",
            _ => null,
        };
        var ptr = NativeMethods.agentstategraph_list_taints(
            _handle.DangerousGetHandle(), pathPrefix, kindStr, includeResolved);
        var raw = Json.ThrowIfError(Strings.ConsumeUtf8(ptr), "list_taints");
        try
        {
            using var doc = JsonDocument.Parse(raw);
            if (doc.RootElement.TryGetProperty("taints", out var arr)
                && arr.ValueKind == JsonValueKind.Array)
            {
                var list = JsonSerializer.Deserialize<List<Taint>>(arr.GetRawText(), Json.Options)
                    ?? new List<Taint>();
                return list;
            }
            throw new AgentStateGraphException("list_taints", $"unexpected response: {raw}");
        }
        catch (JsonException ex)
        {
            throw new AgentStateGraphException("list_taints", $"failed to parse response: {ex.Message}", ex);
        }
    }

    /// <summary>
    /// Checks the taint status for <paramref name="path"/> against
    /// <paramref name="agentId"/> at <paramref name="confidence"/>.
    /// </summary>
    public TaintCheck CheckTaint(string path, string agentId = "", double confidence = 1.0)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(path);
        var ptr = NativeMethods.agentstategraph_check_taint(
            _handle.DangerousGetHandle(), path, agentId ?? string.Empty, confidence);
        var raw = Json.ThrowIfError(Strings.ConsumeUtf8(ptr), "check_taint");
        try
        {
            using var doc = JsonDocument.Parse(raw);
            if (doc.RootElement.TryGetProperty("check", out var check))
            {
                return JsonSerializer.Deserialize<TaintCheck>(check.GetRawText(), Json.Options)
                    ?? throw new AgentStateGraphException("check_taint", "deserialized null TaintCheck");
            }
            throw new AgentStateGraphException("check_taint", $"unexpected response: {raw}");
        }
        catch (JsonException ex)
        {
            throw new AgentStateGraphException("check_taint", $"failed to parse response: {ex.Message}", ex);
        }
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private static string ParseIdEnvelope(string? raw, string operation)
    {
        var json = Json.ThrowIfError(raw, operation);
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.TryGetProperty("id", out var id)
                && id.ValueKind == JsonValueKind.String)
            {
                return id.GetString() ?? string.Empty;
            }
            throw new AgentStateGraphException(operation, $"unexpected response: {json}");
        }
        catch (JsonException ex)
        {
            throw new AgentStateGraphException(operation, $"failed to parse response: {ex.Message}", ex);
        }
    }

    private static string BuildRemoveParams(string name, string reason, string agentId, string? proof)
    {
        // The FFI expects {"name": ..., "reason": ..., "agent_id": ..., "proof"?}
        var obj = new Dictionary<string, object?>
        {
            ["name"] = name,
            ["reason"] = reason,
            ["agent_id"] = agentId,
        };
        if (proof is not null)
        {
            obj["proof"] = proof;
        }
        return JsonSerializer.Serialize(obj, Json.Options);
    }

    private static string MergeWithName(JsonElement baseObj, string name)
    {
        var dict = new Dictionary<string, JsonElement>(StringComparer.Ordinal);
        if (baseObj.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in baseObj.EnumerateObject())
            {
                dict[prop.Name] = prop.Value.Clone();
            }
        }
        // Serialize manually to avoid re-parsing. The FFI tolerates string
        // `name` at the top level alongside agent_id / reason fields.
        using var ms = new System.IO.MemoryStream();
        using (var writer = new Utf8JsonWriter(ms))
        {
            writer.WriteStartObject();
            writer.WriteString("name", name);
            foreach (var (k, v) in dict)
            {
                if (k == "name") continue;
                writer.WritePropertyName(k);
                v.WriteTo(writer);
            }
            writer.WriteEndObject();
        }
        return System.Text.Encoding.UTF8.GetString(ms.ToArray());
    }
}

File: bindings/swift/Sources/AgentStateGraph/Taint.swift
import Foundation
import CAgentStateGraph

/// Kind of protective marker: `taint` / `quarantine` / `watch`.
public enum TaintKind: String, Codable, Sendable {
    case taint, quarantine, watch
}

/// Pre-commit-hook behavior of a taint.
public enum TaintEffect: String, Codable, Sendable {
    case warn, block, review, isolate, advisory
}

/// Advisory severity of a taint.
public enum TaintSeverity: String, Codable, Sendable {
    case low, medium, high, critical
}

/// Threshold direction for a watch: `above` / `below`.
public enum WatchDirection: String, Codable, Sendable {
    case above, below
}

/// The over-the-wire shape of a single taint record.
public struct Taint: Codable, Sendable, Equatable {
    public let id: String
    public let path: String
    public let name: String
    public let kind: TaintKind
    public let effect: TaintEffect
    public let severity: TaintSeverity
    public let reason: String
    public let agentId: String
    public let commitId: String
    public let createdAt: String
    public let expiresAt: String?
    public let resolvedAt: String?
    public let resolvedBy: String?
    public let resolvedReason: String?
    public let resolvedProof: String?
    public let propagate: Bool
    public let metadata: [String: JSONValue]?

    enum CodingKeys: String, CodingKey {
        case id, path, name, kind, effect, severity, reason, propagate, metadata
        case agentId = "agent_id"
        case commitId = "commit_id"
        case createdAt = "created_at"
        case expiresAt = "expires_at"
        case resolvedAt = "resolved_at"
        case resolvedBy = "resolved_by"
        case resolvedReason = "resolved_reason"
        case resolvedProof = "resolved_proof"
    }
}

/// Aggregated status returned by `checkTaint`.
public struct TaintCheck: Codable, Sendable, Equatable {
    public let tainted: Bool
    public let quarantined: Bool
    public let watched: Bool
    public let taints: [Taint]
    public let quarantines: [Taint]
    public let watches: [Taint]
    public let canWrite: Bool
    public let requiredConfidence: Double
    public let authorizedAgents: [String]?
    public let isolated: Bool

    enum CodingKeys: String, CodingKey {
        case tainted, quarantined, watched, taints, quarantines, watches, isolated
        case canWrite = "can_write"
        case requiredConfidence = "required_confidence"
        case authorizedAgents = "authorized_agents"
    }
}

// MARK: - Parameter shapes
//
// Field names match the FFI param parser. Note the FFI reads `expires`
// (RFC3339), not `expires_at`.

/// Input to `taint`.
public struct TaintParams: Codable, Sendable {
    public var name: String
    public var effect: TaintEffect
    public var reason: String
    public var severity: TaintSeverity?
    public var expires: String?
    public var propagate: Bool?
    public var agentId: String

    enum CodingKeys: String, CodingKey {
        case name, effect, reason, severity, expires, propagate
        case agentId = "agent_id"
    }

    public init(
        name: String, effect: TaintEffect, reason: String, severity: TaintSeverity? = nil,
        expires: String? = nil, propagate: Bool? = nil, agentId: String
    ) {
        self.name = name; self.effect = effect; self.reason = reason
        self.severity = severity; self.expires = expires; self.propagate = propagate
        self.agentId = agentId
    }
}

/// Input to `quarantine`.
public struct QuarantineParams: Codable, Sendable {
    public var name: String
    public var reason: String
    public var severity: TaintSeverity?
    public var authorizedAgents: [String]
    public var expires: String?
    public var propagate: Bool?
    public var agentId: String

    enum CodingKeys: String, CodingKey {
        case name, reason, severity, expires, propagate
        case authorizedAgents = "authorized_agents"
        case agentId = "agent_id"
    }

    public init(
        name: String, reason: String, severity: TaintSeverity? = nil,
        authorizedAgents: [String] = [], expires: String? = nil, propagate: Bool? = nil,
        agentId: String
    ) {
        self.name = name; self.reason = reason; self.severity = severity
        self.authorizedAgents = authorizedAgents; self.expires = expires
        self.propagate = propagate; self.agentId = agentId
    }
}

/// Input to `watch`.
public struct WatchParams: Codable, Sendable {
    public var name: String
    public var reason: String
    public var metric: String?
    public var threshold: Double?
    public var direction: WatchDirection?
    public var checkIntervalSecs: UInt64?
    public var expires: String?
    public var severity: TaintSeverity?
    public var propagate: Bool?
    public var agentId: String

    enum CodingKeys: String, CodingKey {
        case name, reason, metric, threshold, direction, expires, severity, propagate
        case checkIntervalSecs = "check_interval_secs"
        case agentId = "agent_id"
    }

    public init(
        name: String, reason: String, metric: String? = nil, threshold: Double? = nil,
        direction: WatchDirection? = nil, checkIntervalSecs: UInt64? = nil, expires: String? = nil,
        severity: TaintSeverity? = nil, propagate: Bool? = nil, agentId: String
    ) {
        self.name = name; self.reason = reason; self.metric = metric; self.threshold = threshold
        self.direction = direction; self.checkIntervalSecs = checkIntervalSecs; self.expires = expires
        self.severity = severity; self.propagate = propagate; self.agentId = agentId
    }
}

/// Input to `untaint` / `unquarantine`. `name` is supplied via the method
/// argument; this carries the rest.
public struct UntaintParams: Codable, Sendable {
    public var reason: String
    public var proof: String?
    public var agentId: String

    enum CodingKeys: String, CodingKey {
        case reason, proof
        case agentId = "agent_id"
    }

    public init(reason: String, proof: String? = nil, agentId: String) {
        self.reason = reason; self.proof = proof; self.agentId = agentId
    }
}

/// Input to `unwatch`. Watches are lightweight, so `reason` is optional.
public struct UnwatchParams: Codable, Sendable {
    public var reason: String?
    public var agentId: String

    enum CodingKeys: String, CodingKey {
        case reason
        case agentId = "agent_id"
    }

    public init(reason: String? = nil, agentId: String) {
        self.reason = reason; self.agentId = agentId
    }
}

// MARK: - Repository extensions

extension AgentStateGraph {
    /// Apply a taint on `path`. Returns the new taint's uuid.
    @discardableResult
    public func taint(_ path: String, params: TaintParams, ref: String = "main") throws -> String {
        try decodeApply(applyTaint(agentstategraph_taint_apply, path, try encodeJSON(params), ref, "taint_apply"))
    }

    /// Remove an active taint by name.
    public func untaint(_ path: String, name: String, params: UntaintParams, ref: String = "main") throws {
        let payload = try mergeName(params, name: name)
        _ = try decodeOK(applyTaint(agentstategraph_taint_remove, path, payload, ref, "taint_remove"))
    }

    /// Apply a quarantine on `path`. Returns the new taint id.
    @discardableResult
    public func quarantine(_ path: String, params: QuarantineParams, ref: String = "main") throws -> String {
        try decodeApply(applyTaint(agentstategraph_quarantine_apply, path, try encodeJSON(params), ref, "quarantine_apply"))
    }

    /// Release an active quarantine by name.
    public func unquarantine(_ path: String, name: String, params: UntaintParams, ref: String = "main") throws {
        let payload = try mergeName(params, name: name)
        _ = try decodeOK(applyTaint(agentstategraph_quarantine_release, path, payload, ref, "quarantine_release"))
    }

    /// Attach a watch to `path`. Returns the new taint id.
    @discardableResult
    public func watch(_ path: String, params: WatchParams, ref: String = "main") throws -> String {
        try decodeApply(applyTaint(agentstategraph_watch_apply, path, try encodeJSON(params), ref, "watch_apply"))
    }

    /// Remove an active watch by name.
    public func unwatch(_ path: String, name: String, params: UnwatchParams, ref: String = "main") throws {
        let payload = try mergeName(params, name: name)
        _ = try decodeOK(applyTaint(agentstategraph_watch_remove, path, payload, ref, "watch_remove"))
    }

    /// Every active taint (or all, if `includeResolved`). Both filters optional.
    public func listTaints(
        pathPrefix: String? = nil, kind: String? = nil, includeResolved: Bool = false
    ) throws -> [Taint] {
        let r = try rawHandle()
        let cPrefix = sgDup(pathPrefix); defer { free(cPrefix) }
        let cKind = sgDup(kind); defer { free(cKind) }
        let raw = try consume(
            agentstategraph_list_taints(r, cPrefix, cKind, includeResolved), "list_taints")
        struct Env: Decodable { let taints: [Taint] }
        return try decodeJSON(raw, as: Env.self).taints
    }

    /// Aggregated taint / quarantine / watch status for `path`.
    public func checkTaint(_ path: String, agentId: String, confidence: Double) throws -> TaintCheck {
        let r = try rawHandle()
        let cPath = sgDup(path); defer { free(cPath) }
        let cAgent = sgDup(agentId); defer { free(cAgent) }
        let raw = try consume(
            agentstategraph_check_taint(r, cPath, cAgent, confidence), "check_taint")
        struct Env: Decodable { let check: TaintCheck? }
        guard let c = try decodeJSON(raw, as: Env.self).check else {
            throw AgentStateGraphError.decode("check_taint: missing check envelope")
        }
        return c
    }

    // MARK: internals

    private func applyTaint(
        _ fn: (UnsafeMutableRawPointer?, UnsafePointer<CChar>?, UnsafePointer<CChar>?, UnsafePointer<CChar>?) -> UnsafeMutablePointer<CChar>?,
        _ path: String, _ paramsJSON: String, _ ref: String, _ op: String
    ) throws -> String {
        let r = try rawHandle()
        let cRef = sgDup(ref); defer { free(cRef) }
        let cPath = sgDup(path); defer { free(cPath) }
        let cParams = sgDup(paramsJSON); defer { free(cParams) }
        return try consume(fn(r, cRef, cPath, cParams), op)
    }

    /// The `{"ok":true,"id":"<uuid>"}` envelope from *_apply.
    private func decodeApply(_ raw: String) throws -> String {
        struct Env: Decodable { let ok: Bool; let id: String? }
        let env = try decodeJSON(raw, as: Env.self)
        guard env.ok, let id = env.id else {
            throw AgentStateGraphError.native(raw)
        }
        return id
    }

    /// The `{"ok":true}` envelope from *_remove / release / unwatch.
    private func decodeOK(_ raw: String) throws {
        struct Env: Decodable { let ok: Bool }
        guard try decodeJSON(raw, as: Env.self).ok else {
            throw AgentStateGraphError.native(raw)
        }
    }

    /// Re-serialize an `Encodable` params payload with an injected `name`,
    /// which the FFI param parser reads from the JSON body.
    private func mergeName<T: Encodable>(_ params: T, name: String) throws -> String {
        let data = try JSONEncoder().encode(params)
        var obj = (try JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        obj["name"] = name
        let merged = try JSONSerialization.data(withJSONObject: obj)
        guard let s = String(data: merged, encoding: .utf8) else {
            throw AgentStateGraphError.decode("merge name: not utf-8")
        }
        return s
    }
}
